File: src/tools/bash.zig
const std = @import("std");
const core = @import("phoenix_core");

const max_bytes: usize = 512 * 1024;

fn errResult(allocator: std.mem.Allocator, comptime fmt: []const u8, args: anytype) anyerror!core.tool.ToolResult {
    const msg = try std.fmt.allocPrint(allocator, fmt, args);
    return .{ .output = msg, .is_error = true };
}

fn invoke(self: *const core.Tool, io: std.Io, args: []const u8, allocator: std.mem.Allocator) anyerror!core.tool.ToolResult {
    _ = self;

    var parsed = std.json.parseFromSlice(std.json.Value, allocator, args, .{}) catch {
        return errResult(allocator, "bash: invalid JSON args: {s}", .{args});
    };
    defer parsed.deinit();
    if (parsed.value != .object) return errResult(allocator, "bash: args must be a JSON object", .{});
    const obj = parsed.value.object;

    const cmd_v = obj.get("command") orelse return errResult(allocator, "bash: missing required field 'command'", .{});
    if (cmd_v != .string) return errResult(allocator, "bash: 'command' must be a string", .{});
    const command = cmd_v.string;

    var timeout_secs: ?i64 = null;
    if (obj.get("timeout")) |v| {
        if (v != .integer or v.integer < 1) return errResult(allocator, "bash: 'timeout' must be a positive integer (seconds)", .{});
        timeout_secs = v.integer;
    }

    const timeout: std.Io.Timeout = if (timeout_secs) |s|
        .{ .duration = .{ .clock = .awake, .raw = .fromSeconds(s) } }
    else
        .none;

    const result = std.process.run(allocator, io, .{
        .argv = &.{ "/bin/sh", "-c", command },
        .stdout_limit = .limited(max_bytes),
        .stderr_limit = .limited(max_bytes),
        .timeout = timeout,
    }) catch |err| {
        if (err == error.Timeout) {
            const msg = try std.fmt.allocPrint(
                allocator,
                "[timed out after {d}s]\nbash: command did not complete before deadline",
                .{timeout_secs.?},
            );
            return .{ .output = msg, .is_error = true };
        }
        return errResult(allocator, "bash: spawn/run failed: {s}", .{@errorName(err)});
    };
    defer allocator.free(result.stdout);
    defer allocator.free(result.stderr);

    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(allocator);

    switch (result.term) {
        .exited => |code| try out.print(allocator, "exit code: {d}\n", .{code}),
        .signal => |sig| try out.print(allocator, "killed by signal: {s}\n", .{@tagName(sig)}),
        .stopped => |sig| try out.print(allocator, "stopped by signal: {s}\n", .{@tagName(sig)}),
        .unknown => |code| try out.print(allocator, "unknown termination: {d}\n", .{code}),
    }

    if (result.stdout.len > 0) {
        try out.appendSlice(allocator, "--- stdout ---\n");
        try out.appendSlice(allocator, result.stdout);
        if (result.stdout[result.stdout.len - 1] != '\n') try out.append(allocator, '\n');
    }
    if (result.stderr.len > 0) {
        try out.appendSlice(allocator, "--- stderr ---\n");
        try out.appendSlice(allocator, result.stderr);
        if (result.stderr[result.stderr.len - 1] != '\n') try out.append(allocator, '\n');
    }

    const owned = try out.toOwnedSlice(allocator);
    const is_error = switch (result.term) {
        .exited => |code| code != 0,
        else => true,
    };
    return .{ .output = owned, .is_error = is_error };
}

pub const bash_tool: core.Tool = .{
    .name = "bash",
    .description = "Execute a bash command in the current working directory. Returns stdout and stderr. Optionally provide a timeout in seconds.",
    .schema =
    \\{"type":"object","properties":{"command":{"type":"string","description":"Bash command to execute"},"timeout":{"type":"integer","minimum":1,"description":"Timeout in seconds (optional, no default timeout)"}},"required":["command"]}
    ,
    .max_output_bytes = max_bytes,
    .invokeFn = invoke,
};

// ---- Tests ----

test "bash captures stdout" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const args =
        \\{"command":"printf hello"}
    ;
    const r = try bash_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "exit code: 0") != null);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "hello") != null);
    try std.testing.expect(!r.is_error);
}

test "bash non-zero exit is error" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const args =
        \\{"command":"exit 7"}
    ;
    const r = try bash_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "exit code: 7") != null);
    try std.testing.expect(r.is_error);
}

test "bash captures stderr" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const args =
        \\{"command":"printf err 1>&2; exit 0"}
    ;
    const r = try bash_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "stderr") != null);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "err") != null);
}

test "bash timeout kills long-running command" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const args =
        \\{"command":"sleep 10","timeout":1}
    ;
    const r = try bash_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "timed out") != null);
    try std.testing.expect(r.is_error);
}

test "bash missing command is error" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const r = try bash_tool.invoke(io, "{}", a);
    defer a.free(r.output);
    try std.testing.expect(r.is_error);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "missing required field 'command'") != null);
}

File: src/tools/bash.rs
use async_trait::async_trait;
use serde_json::{Value, json};
use std::process::Stdio;
use tokio::io::AsyncReadExt;

use super::traits::{InputRequester, Tool, ToolError, ToolResult, ToolSchema};

/// Maximum bytes captured from stdout or stderr.
const MAX_OUTPUT_BYTES: usize = 512 * 1024;

pub struct BashTool;

#[async_trait]
impl Tool for BashTool {
    fn schema(&self) -> ToolSchema {
        ToolSchema {
            name: "bash".into(),
            description: "Execute a bash command in the current working directory. \
                          Returns stdout and stderr. Optionally provide a timeout in seconds."
                .into(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "command": {
                        "type": "string",
                        "description": "Bash command to execute"
                    },
                    "timeout": {
                        "type": "integer",
                        "minimum": 1,
                        "description": "Timeout in seconds (optional, no default timeout)"
                    }
                },
                "required": ["command"]
            }),
        }
    }

    async fn invoke(
        &self,
        args: Value,
        _input: &dyn InputRequester,
    ) -> Result<ToolResult, ToolError> {
        let command = args
            .get("command")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::InvalidArgs("missing required field 'command'".into()))?;

        let timeout_secs = args.get("timeout").and_then(|v| v.as_u64());

        let mut child = tokio::process::Command::new("/bin/sh")
            .arg("-c")
            .arg(command)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .map_err(|e| ToolError::ExecutionFailed(e.to_string()))?;

        let stdout_handle = child.stdout.take();
        let stderr_handle = child.stderr.take();

        // Read stdout and stderr concurrently, capped at MAX_OUTPUT_BYTES.
        let read_capped = |mut reader: Option<tokio::process::ChildStdout>,
                           mut reader_err: Option<tokio::process::ChildStderr>|
         -> (
            tokio::task::JoinHandle<Vec<u8>>,
            tokio::task::JoinHandle<Vec<u8>>,
        ) {
            let out_handle = tokio::spawn(async move {
                let mut buf = vec![0u8; MAX_OUTPUT_BYTES];
                let mut total = 0usize;
                if let Some(ref mut r) = reader {
                    loop {
                        if total >= MAX_OUTPUT_BYTES {
                            break;
                        }
                        match r.read(&mut buf[total..]).await {
                            Ok(0) => break,
                            Ok(n) => total += n,
                            Err(_) => break,
                        }
                    }
                }
                buf.truncate(total);
                buf
            });

            let err_handle = tokio::spawn(async move {
                let mut buf = vec![0u8; MAX_OUTPUT_BYTES];
                let mut total = 0usize;
                if let Some(ref mut r) = reader_err {
                    loop {
                        if total >= MAX_OUTPUT_BYTES {
                            break;
                        }
                        match r.read(&mut buf[total..]).await {
                            Ok(0) => break,
                            Ok(n) => total += n,
                            Err(_) => break,
                        }
                    }
                }
                buf.truncate(total);
                buf
            });

            (out_handle, err_handle)
        };

        let (out_jh, err_jh) = read_capped(stdout_handle, stderr_handle);

        // Wait for exit, optionally with a timeout.
        let status = if let Some(secs) = timeout_secs {
            let deadline = tokio::time::Duration::from_secs(secs);
            match tokio::time::timeout(deadline, child.wait()).await {
                Ok(Ok(s)) => s,
                Ok(Err(e)) => {
                    return Err(ToolError::ExecutionFailed(format!(
                        "failed to wait on child: {e}"
                    )));
                }
                Err(_elapsed) => {
                    // Kill the child on timeout.
                    let _ = child.kill().await;
                    return Err(ToolError::Timeout);
                }
            }
        } else {
            child
                .wait()
                .await
                .map_err(|e| ToolError::ExecutionFailed(format!("failed to wait on child: {e}")))?
        };

        let stdout_bytes = out_jh.await.unwrap_or_default();
        let stderr_bytes = err_jh.await.unwrap_or_default();

        let stdout = String::from_utf8_lossy(&stdout_bytes);
        let stderr = String::from_utf8_lossy(&stderr_bytes);

        let exit_code = status.code().unwrap_or(-1);
        let truncated =
            stdout_bytes.len() >= MAX_OUTPUT_BYTES || stderr_bytes.len() >= MAX_OUTPUT_BYTES;

        // Build combined output.
        let mut out = String::new();
        out.push_str(&format!("exit code: {exit_code}\n"));

        if !stdout.is_empty() {
            out.push_str("--- stdout ---\n");
            out.push_str(&stdout);
            if !stdout.ends_with('\n') {
                out.push('\n');
            }
        }
        if !stderr.is_empty() {
            out.push_str("--- stderr ---\n");
            out.push_str(&stderr);
            if !stderr.ends_with('\n') {
                out.push('\n');
            }
        }

        Ok(ToolResult {
            output: out,
            truncated,
            is_error: exit_code != 0,
        })
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tools::traits::NoopInputRequester;

    #[tokio::test]
    async fn bash_captures_stdout() {
        let tool = BashTool;
        let result = tool
            .invoke(json!({"command": "echo hello"}), &NoopInputRequester)
            .await
            .unwrap();
        assert!(result.output.contains("hello"));
        assert!(result.output.contains("exit code: 0"));
        assert!(!result.is_error);
    }

    #[tokio::test]
    async fn bash_nonzero_exit() {
        let result = BashTool
            .invoke(json!({"command": "exit 42"}), &NoopInputRequester)
            .await
            .unwrap();
        assert!(result.output.contains("exit code: 42"));
        assert!(result.is_error);
    }

    #[tokio::test]
    async fn bash_captures_stderr() {
        let result = BashTool
            .invoke(json!({"command": "echo err >&2"}), &NoopInputRequester)
            .await
            .unwrap();
        assert!(result.output.contains("--- stderr ---"));
        assert!(result.output.contains("err"));
        assert!(!result.is_error);
    }

    #[tokio::test]
    async fn bash_timeout_kills_command() {
        let result = BashTool
            .invoke(
                json!({"command": "sleep 60", "timeout": 1}),
                &NoopInputRequester,
            )
            .await;
        match result {
            Err(ToolError::Timeout) => {} // expected
            other => panic!("expected Timeout error, got: {other:?}"),
        }
    }

    #[tokio::test]
    async fn bash_missing_command() {
        let result = BashTool.invoke(json!({}), &NoopInputRequester).await;
        match result {
            Err(ToolError::InvalidArgs(msg)) => {
                assert!(msg.contains("command"), "message was: {msg}");
            }
            other => panic!("expected InvalidArgs, got: {other:?}"),
        }
    }

    #[tokio::test]
    async fn bash_stdout_and_stderr_combined() {
        let result = BashTool
            .invoke(
                json!({"command": "echo out; echo err >&2"}),
                &NoopInputRequester,
            )
            .await
            .unwrap();
        assert!(result.output.contains("--- stdout ---"));
        assert!(result.output.contains("out"));
        assert!(result.output.contains("--- stderr ---"));
        assert!(result.output.contains("err"));
    }

    #[tokio::test]
    async fn bash_truncated_flag_on_large_output() {
        // Generate just over MAX_OUTPUT_BYTES of stdout.
        // 512KB = 524288 bytes. We'll ask dd to produce exactly that plus some.
        let cmd = format!(
            "dd if=/dev/zero bs=1024 count={} 2>/dev/null | tr '\\0' 'A'",
            (MAX_OUTPUT_BYTES / 1024) + 1
        );
        let result = BashTool
            .invoke(json!({"command": cmd}), &NoopInputRequester)
            .await
            .unwrap();
        assert!(result.truncated);
    }
}
